### PR TITLE
Add basic types for sub-encoder

### PR DESCRIPTION
**Pull request by EvanHahn**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/sub-encoder/tsconfig.json

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5ea166d40c04363e1869b44bff7ec84f2243ceef
</p>

File: types/sub-encoder/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": ["es6"],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
+        "types": [],
```
Comment (codershiba) [inline types/sub-encoder/tsconfig.json:9]:
```suggestion
        "types": ["node"],
```

**Comment by EvanHahn** [comment]
There are several CI failures which seem unrelated to my changes. Apologies if I did something wrong here.

File: types/sub-encoder/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/// <reference types="node" />
+
+type JsonPrimitive = string | number | boolean | null;
```
Comment (codershiba) [inline types/sub-encoder/index.d.ts:3]:
```suggestion
export type JsonPrimitive = string | number | boolean | null;
```

File: types/sub-encoder/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+type JsonArray = JsonValue[];
```
Comment (codershiba) [inline types/sub-encoder/index.d.ts:4]:
```suggestion
export type JsonArray = JsonValue[];
```

File: types/sub-encoder/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+interface JsonObject {
```
Comment (codershiba) [inline types/sub-encoder/index.d.ts:5]:
```suggestion
export interface JsonObject {
```

File: types/sub-encoder/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    [key: string]: JsonValue | undefined;
+}
+type JsonValue = JsonPrimitive | JsonArray | JsonObject;
```
Comment (codershiba) [inline types/sub-encoder/index.d.ts:8]:
```suggestion
export type JsonValue = JsonPrimitive | JsonArray | JsonObject;
```

**Comment by EvanHahn** [comment]
@codershiba Happy to export those types, but why? (Not saying you're wrong, just that I don't know the reason.)

**Comment by codershiba** [comment]
For user consumption. Example could be the the tests itself where you have used `JsonValue` type.

**Comment by EvanHahn** [comment]
Actually, this doesn't work because this module has only one export. Reverted.

File: types/sub-encoder/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    binary: Encoding<string | Buffer, Buffer>;
+    hex: Encoding<string>;
+    json: Encoding<JsonValue>;
+    ndjson: Encoding<JsonValue>;
+    ucs2: Encoding<string>;
+    "utf-8": Encoding<string>;
+    utf16le: Encoding<string>;
+}
+
+declare class SubEncoder<T extends keyof EncodingLookup = "binary"> {
```
[the remark this answers is not in the document]
Comment (EvanHahn) [reply types/sub-encoder/index.d.ts:26]: Done in 5df02ef971f5d02f72cc915c8606844992c29486.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/sub-encoder/sub-encoder-tests.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5df02ef971f5d02f72cc915c8606844992c29486
</p>

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@EvanHahn thanks!

**Comment by EvanHahn** [comment]
Ready to merge

**Pull request merged**